tiny_dnn/core/framework/tensor.h

```diff
@@ -0,0 +1,281 @@
+      and/or other materials provided with the distribution.
+
+    * Neither the name of tiny-dnn nor the names of its
+      contributors may be used to endorse or promote products derived from
+      this software without specific prior written permission.
+
+    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+    AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+    IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+    DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+    FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+    DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+    SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+    OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+    OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+#pragma once
+
+#include "tiny_dnn/core/framework/device.fwd.h"
+
+namespace tiny_dnn {
+
+template<typename U = float_t>
+class Tensor {
+ public:
+    /*
+     * Initializes an empty tensor.
+     */
+    Tensor() {}
+
+    /*
+     * Create a tensor of the given dimension.
+     */
+    explicit Tensor(const cnn_size_t d0,
+                    const cnn_size_t d1,
+                    const cnn_size_t d2,
+                    const cnn_size_t d3) {
+        reshape(d0, d1, d2, d3);
+        resize();
+    }
+
+    // Returns the tensor shape
+    std::vector<cnn_size_t> shape() const { return shape_; }
+    
+    // Returns the value of a specified index in the tensor.
+    // Checked version (throw exceptions for out-of-range error)
+    template<typename T>
+    T& at(const cnn_size_t d0,
+          const cnn_size_t d1,
+          const cnn_size_t d2,
+          const cnn_size_t d3) {
+        return *access_data<T>(d0, d1, d2, d3);
+    }
+
+    template<typename T>
+    const T& at(const cnn_size_t d0,
+                const cnn_size_t d1,
+                const cnn_size_t d2,
+                const cnn_size_t d3) const {
+        return *access_data<T>(d0, d1, d2, d3);
+    }
+
+    // Returns the pointer to a specified index in the tensor
+    // Checked version (throw exceptions for out-of-range error)
+    template<typename T>
+    T* ptr(const cnn_size_t d0,
+           const cnn_size_t d1,
+           const cnn_size_t d2,
+           const cnn_size_t d3) {
+        return access_data<T>(d0, d1, d2, d3);
+    }
+    
+    template<typename T>
+    const T* ptr(const cnn_size_t d0,
+                 const cnn_size_t d1,
+                 const cnn_size_t d2,
+                 const cnn_size_t d3) const {
+        return access_data<T>(d0, d1, d2, d3);
+    }
+
+    // zero-overhead version (same performance to raw pointer access.
+    // have an assertion for out-of-range error)
```

Comment:
In access_data method, I've confirmed vector::at is replaced with operator [ ] but it still uses if statement. If you use non-compile-time constant value in if statement, compiler has to emit branch instruction so it has a little overhead to execute (even in release build). So you should use assert statement (this will be removed in release build) to check index.

Before
```
        if (index >= host_data_->size()) {
            throw nn_error("Access tensor out of range.");
        }
```

After
```
        assert(index < host_data_->size());
```

And there is another if statement.
```
        if (!std::is_same<T,U>::value) {
            return reinterpret_cast<T*>(value);
        }
```

This if statement will probablly be removed in release build. If not, you may need to use C++ technique like template meta programming to optimize it away. Then code will start to get messy like bxxst.